EIPS/eip-BN256_HashToCurve.md

```diff
@@ -0,0 +1,342 @@
+combined into a group signature offline, but there is no
+easy way to verify partial signatures or group signatures
+in the EVM.
+
+In order to perform BLS signature validation, a hash-to-curve
+function is required.
+While the MapToGroup method initially discussed in the original BLS
+[paper](https://crypto.stanford.edu/~dabo/pubs/abstracts/weilsigs.html)
+works in practice, the nondeterministic nature of the algorithm
+leaves something to be desired, as we would like to bound
+the overall computational cost in the EVM.
+A deterministic method for mapping to BN curves is given
+[here](https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf);
+in the paper, Fouque and Tibouchi proved their mapping
+was indifferentiable from a random oracle.
+This paper gives us the desired algorithm.
+
+## Specification
+We mostly follow the Fouque and Tibouchi
+[paper](https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf)
+for the hash-to-curve algorithm.
+This algorithm, as with the current IRTF
+[draft](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10),
+separates the hash-to-curve operation into a
+hash-to-field function and a deterministic field-to-curve map.
+We include some improvements from another
+[paper](https://eprint.iacr.org/2019/403);
+one improvement enables there to be no special points in the
+deterministic mapping as long as the multiplicative inverse
+function call ensures `inverse(0) == 0`;
+this always happens when inversion is computed via exponentiation.
+Furthermore, the Wahby and Boneh paper allows us to make
+a hash-to-curve algorithm for mapping to G2;
+the algorithm is effectively the same, and the only difference
+in the different constants and the base field.
+All of the required operations in these functions
+involve operations on the underlying finite fields.
+Algorithms for necessary GF(p^2) operations are referenced below.
+
+On a local machine, ScalarMult was clocked at 68 microseconds
+per operation.
+The same machine clocked HashToG1 at 94 microseconds per operation
+when hashing 32 bytes into G1; it took 105 microseconds per operation
+when hashing 1024 bytes into G1.
+Given that it currently costs 6000 gas for ECMul, this gives us
+an estimated gas cost for HashToG1 at 8500 + len(bytes).
+Similarly, HashToG2 was clocked at 886 microseconds per operation
+when hashing 32 bytes into G2 and 912 microseconds per operation when
+hashing 1024 bytes into G2.
+This allows us to estimate the gas cost at 80000 + 3*len(bytes).
+
+We now give the pseudocode the HashToG1 function:
+
+```
+function HashToG1(msg)
+    t0 = HashToBase(msg, 0x00, 0x01)
+    t1 = HashToBase(msg, 0x02, 0x03)
+    h0 = BaseToG1(t0)
+    h1 = BaseToG1(t1)
+    h = ECAdd(h0, h1)       # Elliptic curve addition
+    return h
+end function
+```
+
+Here is the pseudocode for HashToBase;
+msg is the byte slice to be hashed and i and j
+are bytes for domain separation.
+p is the prime of the underlying field.
```

Comment:
This is more of a personal recommendation rather than that of an editor, but rather than saying "`i` and `j` are bytes for domain separation", why not just use variable names in the pseudocode that aren't single letters long?  Perhaps just name the variables `domainSeparator1` and `domainSeparator2` or something?  Same with `p`, it seems silly to define it here as "the prime of the underlying field" and then refer to it throughout the code as `p`, why not just name it `fieldPrime` or something in the code so users don't have to use a lookup table?

Same feedback for basically all of the variables in this pseudocode (and really the same feedback for all mathematicians who write code 😬).

Reply:
Algorithms have been updated with variable names, hopefully resulting in pseudocode that is more understandable.